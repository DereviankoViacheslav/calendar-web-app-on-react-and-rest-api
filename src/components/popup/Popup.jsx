import React from 'react';
import './Popup.scss';
import FormCreateEvent from '../form-create-event';
import moment from 'moment';
import PropTypes from 'prop-types';

class Popup extends React.Component {
  state = {
    name: '',
    startDate: moment().format('YYYY-MM-DD'),
    startTime: `${moment().add(1, 'hours').format('HH')}:00`,
    endDate: moment().format('YYYY-MM-DD'),
    endTime: `${moment().add(2, 'hours').format('HH')}:00`,
    description: '',
    color: '#4183f1',
    isShowTrashIcon: false
  }
  
  componentDidMount() {
    const { event, date } = this.props;
    if (date) {
      this.setState({
        startDate: moment(date).format('YYYY-MM-DD'),
        startTime: moment(date).format('HH:mm'),
        endDate: moment(date).format('YYYY-MM-DD'),
        endTime: moment(date).add(1, 'hours').format('HH:mm'),
      });
    }
    if (event) {
      this.setState({
        name: event.name,
        startDate: moment(event.startDate).format('YYYY-MM-DD'),
        startTime: moment(event.startDate).format('HH:mm'),
        endDate: moment(event.endDate).format('YYYY-MM-DD'),
        endTime: moment(event.endDate).format('HH:mm'),
        description: event.description,
        color: event.color,
        isShowTrashIcon: true
      });
    }
  }

  onChange = (event) => {
    this.setState({ [event.target.name]: event.target.value });
  }
  
  onDeleteEvent = () => {
    if (this.props.event) {
      this.props.onDeleteEvent(this.props.event.id)
    }
  }
  
  handleSubmit = (e) => {
    e.preventDefault();
    const { name, color, startDate, startTime, endDate, endTime, description } = this.state;
    const isEndDay = endTime === '00:00';
    const startDateResult = moment(`${startDate}T${startTime}`).format();
    const endDateResult = moment(`${endDate}T${isEndDay ? '23:59' : endTime}`).format();
    const event = {
      createDate: moment().format(),
      startDate: startDateResult,
      endDate: endDateResult,
      description,
      color,
      name,
    };
    if (this.props.event) event.id = this.props.event.id;
    this.props.handleSubmit(event);
  }

  render() {
    return (
      <div className="popup-container">
        <div onClick={this.props.onClosePopup} className="popup-layer" />
        <FormCreateEvent {...this.state}
          onDeleteEvent={this.onDeleteEvent}
          handleSubmit={this.handleSubmit}
          onClosePopup={this.props.onClosePopup}
          onChange={this.onChange}
          />
      </div>
    );
  }
}

Popup.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  onClosePopup: PropTypes.func.isRequired,
  onDeleteEvent: PropTypes.func.isRequired,
  event: PropTypes.object,
  date: PropTypes.string,
}

Popup.defaultProps = {
  date: '',
}

export default Popup;