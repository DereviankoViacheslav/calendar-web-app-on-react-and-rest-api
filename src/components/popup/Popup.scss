.popup-layer {
  height: 100vh;
  position: fixed;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  z-index: 999;
  background-color: #89898961;
}