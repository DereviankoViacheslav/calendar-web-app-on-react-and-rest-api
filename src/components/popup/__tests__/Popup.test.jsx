import React from 'react';
import { shallow } from 'enzyme';
import Popup from '../Popup'
import moment from 'moment';

describe('<Popup />', () => {
  const state = {
    startDate: moment('2020-02-12').format('YYYY-MM-DD'),
    startTime: `${moment('2020-02-12').add(1, 'hours').format('HH')}:00`,
    endDate: moment('2020-02-12').format('YYYY-MM-DD'),
    endTime: `${moment('2020-02-12').add(2, 'hours').format('HH')}:00`,
    description: '',
    color: '#4183f1',
    isShowTrashIcon: false
  };
  const props = {
    handleSubmit: jest.fn(),
    onClosePopup: jest.fn(),
    onDeleteEvent: jest.fn(),
    event: {},
    date: '',
  };

  it('Get snapshot', () => {
    const wrappedComponent = shallow(<Popup {...props} />);
    wrappedComponent.setState(state);
    expect(wrappedComponent).toMatchSnapshot();
  });

  it('should render FormCreateEvent', () => {
    const wrappedComponent = shallow(<Popup {...props} />);
    wrappedComponent.setState(state);
    expect(wrappedComponent.find('FormCreateEvent').exists()).toBeTruthy();
  });
});