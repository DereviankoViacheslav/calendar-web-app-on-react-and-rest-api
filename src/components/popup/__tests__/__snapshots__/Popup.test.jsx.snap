// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popup /> Get snapshot 1`] = `
<div
  className="popup-container"
>
  <div
    className="popup-layer"
    onClick={[MockFunction]}
  />
  <FormCreateEvent
    color="#4183f1"
    description=""
    endDate="2020-02-12"
    endTime="02:00"
    handleSubmit={[Function]}
    isShowTrashIcon={false}
    name=""
    onChange={[Function]}
    onClosePopup={[MockFunction]}
    onDeleteEvent={[Function]}
    startDate="2020-02-12"
    startTime="01:00"
  />
</div>
`;
